`RoamingSupported` for `Group`

StageGuard opened the pull request:
已经写好了，但是 test 和 mock 部分还没完成。
通过 `MessageSvc.PbGetGroupMsg` 指令传递 msgSeq 获取消息，然后通过判断 `MsgComm.Msg.msgHead.msgTime` 来截取需要的消息。

StageGuard committed:
[core] decode `MessageSvc.PbGetGroupMsg` and `OidbSvc.0x88d_0`

StageGuard committed:
[core] Proposal implementation of `RoamingSupported` for `Group`

Him188 commented on mirai-core/src/commonMain/kotlin/contact/roaming/RoamingMessagesImpl.kt:
```diff
@@ -106,4 +108,71 @@ internal class RoamingMessagesImplFriend(
+        timeStart: Long,
+        lastMessageTime: Long,
+        random: Long // unused field
+    ): MessageSvcPbGetRoamMsgReq.Response {
```
这个实现复杂度太高了, sequence 每前进一块都要重新从最新的消息开始查

Him188 replied:
我实现了一个更高效的

Him188 committed:
[core] Rename CommonRoamingMessagesImpl to TimeBasedRoamingMessagesImpl. Extract RoamingMessagesImplFriend

Him188 committed:
[core] Add default implementation for `RoamingMessages.getMessagesStream`

Him188 committed:
[core] Implement a more efficient algorithm to fetch roaming messages for group:

- Added `RoamingMessagesImplGroup`.
- Dump API changes for Group RoamingMessages.
- [mock] Fix MockRoamingMessages missing MessageSource
- [core] Convert hierarchical TimeBasedRoamingMessagesImpl to common, to reduce code complexity

Karlatemp commented on mirai-core-mock/test/mock/MessagingTest.kt:
```diff
@@ -148,6 +145,24 @@ internal class MessagingTest: MockBotTestBase() {
+            mockGroupMember3 says { append("msg3") }
+        }
+
+        with(mockGroup.roamingMessages.getAllMessages().toList()) {
+            assertEquals(3, size)
+            assertEquals(messageChainOf(PlainText("msg1")), get(0))
+            assertEquals(messageChainOf(PlainText("msg2")), get(1))
+            assertEquals(messageChainOf(PlainText("msg3")), get(2))
+        }
```
source 呢

Karlatemp commented on mirai-core/src/commonMain/kotlin/contact/roaming/RoamingMessagesImplGroup.kt:
```diff
@@ -0,0 +1,92 @@
+                    )
+                )
+
+                if (resp is MessageSvcPbGetGroupMsg.Failed) break
```
may log warning on failed?

Him188 replied:
不需要，不知道何时会fail，可能fail也是正常情况

Karlatemp commented on mirai-core/src/jvmBaseMain/kotlin/contact/roaming/TimeBasedRoamingMessagesImpl.kt:
```diff
@@ -17,7 +17,7 @@ import net.mamoe.mirai.utils.JavaFriendlyAPI
 import net.mamoe.mirai.utils.stream
 import java.util.stream.Stream
 
-internal actual sealed class RoamingMessagesImpl : CommonRoamingMessagesImpl() {
+internal actual sealed class TimeBasedRoamingMessagesImpl : CommonTimeBasedRoamingMessagesImpl() {
     @JavaFriendlyAPI
     override suspend fun getMessagesStream(
         timeStart: Long,
```
可以把这个 actual 删掉了, 统一使用 flow 减少维护难度

Him188 replied:
确实
